Convert `struct FromBytesWithNulError` into enum

nyurik opened the pull request:
This PR renames the former `kind` enum from `FromBytesWithNulErrorKind` to `FromBytesWithNulError`, and removes the original struct.

See https://github.com/rust-lang/libs-team/issues/493

## Possible Changes - TBD
* [x] should the new `enum FromBytesWithNulError` derive `Copy`?
* [ ] should there be any new/changed attributes?
* [x] add some more tests

## Problem

One of `CStr` constructors, `CStr::from_bytes_with_nul(bytes: &[u8])` handles 3 cases:
1. `bytes` has one NULL as the last value - creates CStr
2. `bytes` has no NULL - error
3. `bytes` has a NULL in some other position - error

The 3rd case is error that may require lossy conversion, but the 2nd case can easily be handled by the user code. Unfortunately, this function returns an opaque `FromBytesWithNulError` error in both 2nd and 3rd case, so the user cannot detect just the 2nd case - having to re-implement the entire function and bring in the `memchr` dependency.

## Motivating examples or use cases

In [this code](https://github.com/gquintard/varnish-rs/blob/f86d7a87683b08d2e634d63e77d9dc1d24ed4a13/varnish-sys/src/vcl/ws.rs#L158), my FFI code needs to copy user's `&[u8]` into a C-allocated memory blob in a NUL-terminated `CStr` format.  My code must first validate if `&[u8]` has a trailing NUL (case 1), no NUL (adds one on the fly - case 2), or NUL in the middle (3rd case - error). I had to re-implement `from_bytes_with_nul` and add `memchr`dependency just to handle the 2nd case.

r? @Amanieu

nyurik committed:
Convert `struct FromBytesWithNulError` into enum

One of `CStr` constructors, `CStr::from_bytes_with_nul(bytes: &[u8])` handles 3 cases:
1. `bytes` has one NULL as the last value - creates CStr
2. `bytes` has no NULL - error
3. `bytes` has a NULL in some other position - error

The 3rd case is error that may require lossy conversion, but the 2nd case can easily be handled by the user code. Unfortunately, this function returns an opaque `FromBytesWithNulError` error in both 2nd and 3rd case, so the user cannot detect just the 2nd case - having to re-implement the entire function and bring in the `memchr` dependency.

In [this code](https://github.com/gquintard/varnish-rs/blob/f86d7a87683b08d2e634d63e77d9dc1d24ed4a13/varnish-sys/src/vcl/ws.rs#L158), my FFI code needs to copy user's `&[u8]` into a C-allocated memory blob in a NUL-terminated `CStr` format.  My code must first validate if `&[u8]` has a trailing NUL (case 1), no NUL (adds one on the fly - case 2), or NUL in the middle (3rd case - error). I had to re-implement `from_bytes_with_nul` and add `memchr`dependency just to handle the 2nd case.

This PR renames the former `kind` enum from `FromBytesWithNulErrorKind` to `FromBytesWithNulError`, and removes the original struct.

nyurik commented:
@tgross35 I thought this proposal has already passed - see https://github.com/rust-lang/libs-team/issues/493

tgross35 commented:
The meeting was a preliminary thumbs up, but the actual FCP process still needs to happen for anything that changes stable API.

nyurik commented:
@tgross35 thx, any links on anything I need to do?  Or is this rust-team internal?

tgross35 commented:
Nothing internal and nothing for you to do, a lot of rust-lang just slows down around this time of year. Somebody from the libs-api team will need to propose FCP, Amanieu can do that and is already assigned.

Amanieu commented on library/core/src/ffi/c_str.rs:
```diff
@@ -126,35 +126,20 @@ pub struct CStr {
-enum FromBytesWithNulErrorKind {
+pub enum FromBytesWithNulError {
+    /// Data provided contains an interior nul byte at byte pos `usize`.
     InteriorNul(usize),
```
This should be changed to a name field with its own doc comment.

nyurik replied:
@Amanieu could you give an example of some other enum that already does this?  I looked at it seems most enums use tuple values.  Also, what do you mean by its own doc comment?  Would this be sufficient?

```rust
/// Data provided contains an interior nul byte at byte `pos`.
InteriorNul { pos: usize },
```

nyurik replied:
ah, nvm about the docs.  I amended my PR - see if that's what you meant

Amanieu commented on library/core/src/ffi/c_str.rs:
````diff
@@ -126,35 +126,20 @@ pub struct CStr {
 /// ```
 #[derive(Clone, PartialEq, Eq, Debug)]
````
Since all fields are exposed then it's fine to make this `Copy`.

nyurik replied:
thx, added

Amanieu commented:
This is insta-stable since this is turned into a `enum` which can be matched on, even if (I think) we put the variants under a feature flag.

@rfcbot merge

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12507732387/job/34894732326) [(plain)](https://github.com/rust-lang/rust/commit/890be42580e80fef7bc4be8edc7cf4bde7152de8/checks/34894732326/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 27.4s done
#22 DONE 32.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling compiler_builtins v0.1.138
error[E0164]: expected tuple struct or tuple variant, found struct variant `Self::InteriorNul`
   --> library/core/src/ffi/c_str.rs:141:13
    |
141 |             Self::InteriorNul(..) => "data provided contains an interior nul byte",
    |             ^^^^^^^^^^^^^^^^^^^^^ not a tuple struct or tuple variant
help: the struct variant's field is being ignored
    |
    |
141 |             Self::InteriorNul { pos: _ } => "data provided contains an interior nul byte",

error[E0164]: expected tuple struct or tuple variant, found struct variant `Self::InteriorNul`
   --> library/core/src/ffi/c_str.rs:187:16
    |
    |
187 |         if let Self::InteriorNul(pos) = self {
    |                ^^^^^^^^^^^^^^^^^^^^^^ not a tuple struct or tuple variant
help: the struct variant's field is being ignored
    |
    |
187 |         if let Self::InteriorNul { pos: _ } = self {

error[E0533]: expected value, found struct variant `FromBytesWithNulError::InteriorNul`
   --> library/core/src/ffi/c_str.rs:367:34
    |
    |
367 |             Some(nul_pos) => Err(FromBytesWithNulError::InteriorNul(nul_pos)),
    |
help: you might have meant to create a new value of the struct
    |
    |
367 |             Some(nul_pos) => Err(FromBytesWithNulError::InteriorNul { pos: /* value */ }),

Some errors have detailed explanations: E0164, E0533.
For more information about an error, try `rustc --explain E0164`.
error: could not compile `core` (lib) due to 3 previous errors
```

</details>

theemathas commented:
I believe that this can potentially break existing code, so this needs at the very least a crater run. See https://github.com/rust-lang/rfcs/pull/3753

tgross35 commented:
@bors try

bors commented:
:hourglass: Trying commit bf4ececd9cf0b45fc95505ce59d7cdcd8285a7e1 with merge aacf02ff7319d9b9af3fe5e9e1094c2f000a4689...
<!-- homu: {"type":"TryBuildStarted","head_sha":"bf4ececd9cf0b45fc95505ce59d7cdcd8285a7e1","merge_sha":"aacf02ff7319d9b9af3fe5e9e1094c2f000a4689"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12617986133/job/35162326811)
Build commit: aacf02ff7319d9b9af3fe5e9e1094c2f000a4689 (`aacf02ff7319d9b9af3fe5e9e1094c2f000a4689`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12617986133/job/35162326811"},"merge_sha":"aacf02ff7319d9b9af3fe5e9e1094c2f000a4689"} -->

tgross35 commented:
I was about to start a crater run but checked for use in the wild first, from a quick search I don't see any cases of the potentially breaking pattern [mentioned in the ACP](https://github.com/rust-lang/libs-team/issues/493#issuecomment-2532785989) https://github.com/search?q=lang%3Arust+%2FFromBytesWithNulError+%3F%5C%7B+%3F%5C.%5C.%2F&type=code. So the beta crater runs are probably sufficient to cover anything here, unless there are other patterns that also break.

SUPERCILEX commented:
Thoughts on the name `MissingNulTerminator` instead of `NotNulTerminated`? I feel like it's a little clearer.

SUPERCILEX commented:
Also `pos` should be `position` since it's [not shortened in other APIs](https://doc.rust-lang.org/std/?search=pos) or maybe even `index`. Not sure between `position` and `index`. And actually maybe it should be `valid_up_to` to be consistent? https://doc.rust-lang.org/std/str/struct.Utf8Error.html#method.valid_up_to

SUPERCILEX commented:
Sorry for the spam, I really should have gathered my thoughts and posted a single comment.

With `Utf8Error` in mind, I'm not sure an enum is the right choice. Or at least the fact that it's insta-stable is scary. I could see a `fn valid_up_to(&self) -> Option<usize>` API making sense. Or if the enum is the right choice, we could add an unsable method that returns an unstable enum for trial purposes and then convert the `FromBytesWithNulError` struct to the enum later if we think that's the right choice.

nyurik committed:
Rename `pos` to `position`

nyurik commented:
Thx for feedback - I renamed it to `position`. I also rebased this PR.

Note that printing this error still shows `pos`:

```rust
write!(f, " at byte pos {position}")?;
```

I think we may want to keep it this way to be consistent with the existing code - in case someone was using error message parsing.

dtolnay reviewed (approved):
Thank you!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 86b86fa8fb89aca91aa176a0727b9f233cd14353 has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 86b86fa8fb89aca91aa176a0727b9f233cd14353 -->
<!-- homu: {"type":"Approved","sha":"86b86fa8fb89aca91aa176a0727b9f233cd14353","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
